### PR TITLE
[SPARK-33753] [CORE] Reduce the memory footprint and gc of the cache (hadoopJobMetadata)

### PR DESCRIPTION
**Pull request by mcdull-zhang**
### What changes were proposed in this pull request?
Set the maximum capacity of the cache to prevent the cache from taking up too much memory.

### Why are the changes needed?
The online job found that the cache used a lot of memory, resulting in serious full gc.
screenshot of memory analyzer:
image

image

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
manual test

### CODE REVIEW

**Comment by mcdull-zhang** [comment]
@cloud-fan  friendly ping

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -1098,6 +1098,15 @@ package object config {
       .booleanConf
       .createWithDefault(true)
 
+  private[spark] val HADOOP_RDD_METADATA_CACHE_SIZE =
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1101]: Let's not make this configurable. Just choose some largeish reasonable cap

Comment (mcdull-zhang) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: It has been modified, and it feels that 1000 is completely enough.

By the way, the size of a jobconf is 110KB, and 1000 pieces will occupy about 110MB.
do you think it's reasonable？

**Review by srowen — APPROVED** [review]
It's good to limit this, and 1000 seems like a reasonable cap. We could raise it later if there were evidence this is ever hit.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Review by dongjoon-hyun — COMMENTED** [review]
+1, LGTM. Thank you, @mcdull-zhang , @srowen , @cloud-fan .

Could you revise the PR description, @mcdull-zhang ? The current one is a little outdated.

> Added an option to control the size of the cache to prevent the cache from taking up too much memory

**Comment by mcdull-zhang** [comment]
> +1, LGTM. Thank you, @mcdull-zhang , @srowen , @cloud-fan .
> 
> Could you revise the PR description, @mcdull-zhang ? The current one is a little outdated.
> 
> > Added an option to control the size of the cache to prevent the cache from taking up too much memory

Modified, Thank you for your review

**Pull request closed without merging**